Fix Android certificate validation JNI calling convention

StefanoDuo opened the pull request:
Use JNI parameter ordering defined for instance methods (instead of the
static methods one).
Add tests to avoid regressions.

#1575

StefanoDuo commented:
/retest

StefanoDuo commented:
@danzh2010 Would you like to review this yourself or should I send this to someone else?

danzh2010 commented on library/common/jni/jni_interface.cc:
```diff
@@ -1146,7 +1145,7 @@ envoy_cert_verify_status_t jvm_cert_get_status(JNIEnv* env, jobject j_result) {
   jmethodID jmid_getStatus = env->GetMethodID(jcls_AndroidCertVerifyResult, "getStatus", "()I");
   envoy_cert_verify_status_t result = static_cast<envoy_cert_verify_status_t>(
-      env->CallIntMethod(jcls_AndroidCertVerifyResult, jmid_getStatus, j_result));
+      env->CallIntMethod(j_result, jmid_getStatus, j_result));
```
Is the 3rd argument needed?

StefanoDuo replied:
It's not, thanks for catching this!

danzh2010 commented on library/java/org/chromium/net/AndroidCertVerifyResult.java:
```diff
@@ -14,7 +14,7 @@ public final class AndroidCertVerifyResult {
    * The verification status. One of the values in CertVerifyStatusAndroid.
    */
-  private final int mStatus;
+  @CertVerifyStatusAndroid private final int mStatus;
```
To educate myself, what is `@CertVerifyStatusAndroid` annotation for?

StefanoDuo replied:
The `@CertVerifyStatusAndroid` annotation is declared at [java/org/chromium/net/CertVerifyStatusAndroid.java](https://sourcegraph.com/github.com/envoyproxy/envoy-mobile/-/blob/library/java/org/chromium/net/CertVerifyStatusAndroid.java).

It basically makes use of Android's [IntDef](https://developer.android.com/reference/androidx/annotation/IntDef) annotation to say: this is a "special int" and it's values should be constrained to those declared by `@CertVerifyStatusAndroid`.

I've taken this approach from Chromium's code, I think it makes sense in this case (compared for example to using an enum) because we will need an actual int to pass it through the JNI layer.

danzh2010 commented on test/java/org/chromium/net/CertificateVerificationTest.java:
```diff
@@ -143,4 +145,24 @@ public void testChainWithMultipleRootCertificates() throws Exception {
     assertEquals(result.getStatus(), CertVerifyStatusAndroid.OK);
   }
+
+  @Test
```
Nice tests! 

It's still unclear to me how the previous tests didn't run into any issue with the incorrect jobject being passed in. Can you explain a bit?

StefanoDuo replied:
I was also confused, but if you look closely at the tests above they don't exercise the path where C++ unpacks AndroidCertVerifyResult through JNI. They just construct the AndroidCertVerifyResult through JNI (`JniLibrary.callCertificateVerificationFromNative`) and then unpack that directly in Java, skipping the wrong JNI code...

The new tests generate the AndroidCertVerifyResult directly in Java, then unpack it through JNI calls, so now we should be covering the entirety of the JNI call surface.

danzh2010 replied:
Wow, that's tricky. But nice that we found it out!

StefanoDuo commented:
/retest

danzh2010 reviewed (dismissed):
LGTM

danzh2010 commented:
/assign @RyanTheOptimist

RyanTheOptimist reviewed (approved):
Thanks for doing this!

StefanoDuo commented on library/java/org/chromium/net/AndroidCertVerifyResult.java:
```diff
@@ -26,29 +26,35 @@ public final class AndroidCertVerifyResult {
+  /**
+   * Called from native.
+   */
+  @CertVerifyStatusAndroid
```
Syntactically both work. I've personally always seen annotation been put before public (e.g. https://developer.android.com/studio/write/annotations#adding-nullness).

Having said that, if you prefer the other way around (or envoy-mobile styleguide requires that) I can move it. Feel free to reopen if that's the case.

StefanoDuo commented on library/java/org/chromium/net/AndroidCertVerifyResult.java:
```diff
@@ -26,29 +26,35 @@ public final class AndroidCertVerifyResult {
+  @CertVerifyStatusAndroid
+  public int getStatus() {
+    return mStatus;
+  }
```
That's what I've discovered:
* If CertVerifyStatusAndroid is before public, that's what we get (not a one liner)
* if CertVerifyStatusAndroid is moved before int, this stays a one liner

Augustyniak commented on library/java/io/envoyproxy/envoymobile/engine/JniLibrary.java:
```diff
@@ -378,7 +378,23 @@ public static native Object callCertificateVerificationFromNative(byte[][] certC
+   *
+   * @param result AndroidCertVerifyResult to call getStatus onto.
+   */
+  public static native int callAndroidCertVerifyResultGetStatusFromNative(Object result);
```
If this is supposed to be called from within tests only can we make it a part of a separate JNI interface - something such as `JNITestLibrary` or similar and add these 2 methods in there (maybe we could make JNITestLibrary to inherit from `JNILibrary`)? We could then include `JNILibrary` and these 2 methods in a test target only. That would allow us to avoid leaking testing utility methods in production code (and prevent accidental prod use of these methods).

StefanoDuo replied:
I've been trying to split the JNI library, but I've been hitting a bunch of Java runtime errors (I'm not entirely sure if I'm supposed to register the additional testing library somewhere?).

How strongly do you feel about this? As in, I completely agree that it's a good thing to do, but it's also relatively obvious that these are methods to be used only for testing. So, I'm not sure how much time is worth spending into this to get it "right/bullet proof".

Augustyniak reviewed (commented):
Looks good, thank you for fixing these! let one comment

I've seen a comment related to this already but just to reiterate - we will need to change `env->FindClass` to `find_class` to make this work on an actual device I think.

RyanTheOptimist commented:
@StefanoDuo is this ready to merge?

RyanTheOptimist commented:
/wait

RyanTheOptimist commented:
/wait

StefanoDuo commented:
/retest

StefanoDuo commented:
P.S. Sorry for the late follow-up. I've been OOO for the past month.

AFAIU, most of these CL has been merged by @danzh2010 through https://github.com/envoyproxy/envoy-mobile/pull/2554. What's missing is just these additional tests, that prevent the calling convention bug to resurface in the future.

I've rebased and updated the code to reflect that.

jpsim commented:
Closing this PR as this repo is now archived. You can open a new PR in the Envoy repo if you'd like: https://github.com/envoyproxy/envoy
